File: BinarySearch/binsearch.lua
function binSearch(arr, find)
    local start, fin = 1, #arr
    local mid = math.floor((start + fin) / 2)

    while start <= fin do
        if find == arr[mid] then
            return mid
        elseif find < arr[mid] then
            fin = mid - 1
            mid = math.floor((start + fin) / 2)
        else
            start = mid + 1
            mid = math.floor((start + fin) / 2)
        end
    end

    return -1
end

function main()
    io.write("Enter number of elements: ")
    local n = io.read("*n")
    local arr = {}

    for i = 1, n do
        io.write("Enter an element: ")
        arr[i] = io.read("*n")
    end

    io.write("Enter an element to search: ")
    local find = io.read("*n")

    table.sort(arr)

    print("Sorted array: ")

    for i = 1, n do
        io.write(arr[i].." ")
    end
    print()

    local ans = binSearch(arr, find)

    if ans == -1 then
        print("Element not found")
    else
        print("Element found at position: "..ans)
    end
end

main()

File: BinarySearch/binsearch.cpp
#include <iostream>

#define cin std::cin
#define cout std::cout

int binarySearch(int arr[], int n, int find) {
    int start = 0, end = n - 1;
    int mid = (start + end) / 2;

    while(start <= end) {
        if(find == arr[mid]) {
            return mid + 1;
        }
        else if(find < arr[mid]) {
            end = mid - 1;
            mid = (start + end) / 2;
        }
        else {
            start = mid + 1;
            mid = (start + end) / 2;
        }
    }
    return -1;
}

int main() {
    int n;
    cout << "Enter number of elements: ";
    cin >> n;
    int arr[n], find;

    for(int i = 0; i < n; i++) {
        cout << "Enter an element: ";
        cin >> arr[i];
    }

    cout << "Enter an element to search: ";
    cin >> find;

    int ans = binarySearch(arr, n, find);

    if(ans == -1) {
        cout << "Element not found";
    }
    else {
        cout << "Element found at position: " << ans;
    }
}
